### PR TITLE
cleanup(generator): generalize return type strings

### PR DESCRIPTION
**Pull request by cuiy0006**
Generalize return type strings of the non-lro methods in generator.

Only clean up for 2 generators: 
generator/internal/auth_decorator_generator.cc
generator/internal/client_generator.cc

There are more cleanups to do in the followup PR:
generator/internal/connection_generator.cc
generator/internal/connection_impl_generator.cc
generator/internal/connection_impl_rest_generator.cc
generator/internal/logging_decorator_generator.cc
generator/internal/logging_decorator_rest_generator.cc
generator/internal/metadata_decorator_generator.cc
generator/internal/metadata_decorator_rest_generator.cc
generator/internal/mock_connection_generator.cc
generator/internal/round_robin_decorator_generator.cc
generator/internal/stub_generator_base.cc
generator/internal/stub_generator.cc
generator/internal/stub_rest_generator.cc
generator/internal/tracing_connection_generator.cc
generator/internal/tracing_stub_generator.cc

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14657)

### CODE REVIEW

**Review by dbolduc — CHANGES_REQUESTED** [review]
Good idea to stop after 1 or 2 generators.

File: generator/internal/auth_decorator_generator.cc [hunk 1]
```diff
@@ -217,8 +218,9 @@ std::unique_ptr<google::cloud::internal::StreamingReadRpc<$response_type$>>
       continue;
     }
     CcPrintMethod(method, __FILE__, __LINE__,
-                  IsResponseTypeEmpty(method) ? "\nStatus"
-                                              : "\nStatusOr<$response_type$>");
+                  FormatMethodReturnType(method, /*is_async=*/false,
```
[the remark this answers is not in the document]
Comment (cuiy0006) [reply generator/internal/auth_decorator_generator.cc:221]: I agree with using `method_vars`, that is better than my helper function.

But I think `return_type` is not enough, one method can be generated as sync and async versions at the same time. We may need both `sync_return_type` and `async_return_type`.

File: generator/internal/descriptor_utils.cc [hunk 2]
```diff
@@ -874,6 +875,36 @@ Status PrintMethod(google::protobuf::MethodDescriptor const& method,
   return {};
 }
 
+std::string FormatMethodReturnType(
+    google::protobuf::MethodDescriptor const& method, bool is_async,
+    bool is_longrunning, std::string const& prefix, std::string const& suffix) {
+  bool is_response_type_empty = IsResponseTypeEmpty(method);
+  std::string return_type;
+
+  if (is_async) {
```
[the remark this answers is not in the document]
Comment (cuiy0006) [reply generator/internal/descriptor_utils.cc:884]: There is an exception. For longrunning, the sync return type includes `longrunning_operation_type` while the async version is `longrunning_deduced_response_type`.

File: generator/internal/auth_decorator_generator.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (cuiy0006) [reply generator/internal/auth_decorator_generator.cc:221]: Using `method_vars` is a good idea, it's better than the helper function.

But as the comment above, longrunning response types are different between sync and async versions. So I think we may need both `sync_ return_type` and `async_return_type`.

File: generator/internal/descriptor_utils.h [hunk 3]
```diff
@@ -93,6 +93,17 @@ bool CheckParameterCommentSubstitutions();
...
 std::string CppTypeToString(google::protobuf::FieldDescriptor const* field);
 
+/**
+ * Format the return type of methods by
+ * 1. is method return type empty
+ * 2. is method async
+ * 3. is method longrunning
+ * 4. prefix and suffix
+ */
+std::string FormatMethodReturnType(
```
[the remark this answers is not in the document]
Comment (cuiy0006) [reply generator/internal/descriptor_utils.h:103]: Yeah, if we use `method_vars`, the function makes no sense, I will remove it.

File: generator/internal/descriptor_utils.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbolduc) [reply generator/internal/descriptor_utils.cc:884]: I am not suggesting that we change the LRO code to use `$return_type$`.

[the remark this answers is not in the document]
Comment (cuiy0006) [reply generator/internal/descriptor_utils.cc:884]: After discussion with @dbolduc, we decide to only generalize the return string for non-longrunning types.

File: generator/internal/auth_decorator_generator.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (cuiy0006) [reply generator/internal/auth_decorator_generator.cc]: I will only generalize non-lro return type.

**Review by dbolduc — COMMENTED** [review]
Can you add a simple test case to `descriptor_utils_test.cc` to verify the `Status` and `StatusOr<T>` cases of `return_type`?

Something like this:

https://github.com/googleapis/google-cloud-cpp/blob/0e6f85404a95c4de085ae6f905216505f25fe32a/generator/internal/descriptor_utils_test.cc#L1258-L1259

**Review by dbolduc — APPROVED** [review]

**Pull request merged**